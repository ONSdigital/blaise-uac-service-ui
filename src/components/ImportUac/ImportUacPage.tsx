import React, { ReactElement, useState } from "react";
import { Formik, Form } from "formik";
import { Link, Route, Routes } from "react-router-dom";
import App from "../../App";
import SelectFile from "./Sections/SelectFile";
import { ONSButton } from "blaise-design-system-react-components";
import { importUacsFromFile } from "../../client/file-functions";
import UploadFailed from "./Sections/UploadFailed";
import UploadSuccessful from "./Sections/UploadSuccessful";
import { AxiosError } from "axios";

enum Step {
  SelectFile,
  UploadFailed,
  UploadSuccessful
}

function ImportUacPage(): ReactElement {
    const [file, setFile] = useState<File>();
    const [error, setError] = useState<Error | AxiosError>();
    const [uacsImported, setUacsImported] = useState<number>();
    const [activeStep, setActiveStep] = useState(Step.SelectFile);

    function _renderStepContent(step: number) {
        switch (step) {
        case Step.SelectFile:
            return (<SelectFile file={file} setFile={setFile}/>);
        case Step.UploadFailed:
            return (<UploadFailed error={error} />);
        case Step.UploadSuccessful:
            return (<UploadSuccessful uacsImported={uacsImported} />);
        }
    }

    async function _handleSubmit() {
        switch (activeStep) {
        case Step.SelectFile: {
            await importUacsFromFile(file).then((uacsImported: number) => {
                setUacsImported(uacsImported);
                setActiveStep(Step.UploadSuccessful);
            }).catch((error: Error) => {
                setError(error);
                setActiveStep(Step.UploadFailed);
            });
            break;
        }
        default:
            setActiveStep(Step.SelectFile);
        }
    }

    return (
        <>
            <main id="main-content" className="ons-page__main ons-u-mt-no">
                <Routes>
                    <Route path="/app">
                        <App />
                    </Route>
                    <Route path="/">

                        <ul className="ons-list ons-list--bare ons-list--inline ons-u-mt-m">
                            <li className="ons-list__item">
                                <Link to="/app" id="generate-uac-link">
                  Home
                                </Link>
                            </li>
                        </ul>
                    </Route>
                </Routes>

                <Formik
                    validateOnBlur={false}
                    validateOnChange={false}
                    initialValues={{ }}
                    onSubmit={_handleSubmit}>
                    {({ isSubmitting }) => (
                        <Form id={"formID"}>
                            <div>
                                <div className="ons-u-mt-m">
                                    {_renderStepContent(activeStep)}
                                </div>
                                <div className="ons-btn-group ons-u-mt-m">
                                    <ONSButton
                                        id={"next-screen-button"}
                                        submit={true}
                                        loading={isSubmitting}
                                        primary={true}
                                        label={"Continue"}
                                        onClick={(e) => e.currentTarget.blur()} /></div>
                            </div>
                        </Form>
                    )}
                </Formik>
            </main>
        </>
    );
}

export default ImportUacPage;
