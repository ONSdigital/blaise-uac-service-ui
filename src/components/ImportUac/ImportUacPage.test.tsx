/**
 * @jest-environment jsdom
 */

import React from "react";
import { Router } from "react-router";
import { render, waitFor, fireEvent, cleanup, screen, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { createMemoryHistory } from "history";
import ImportUacPage from "./ImportUacPage";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";

const mock = new MockAdapter(axios, { onNoMatch: "throwException" });

describe("Import UAC Page", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
  });

  it("import uac page matches Snapshot", async () => {
    const history = createMemoryHistory();
    const wrapper = render(
      <Router history={history}>
        <ImportUacPage />
      </Router>
    );

    await act(async () => await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    }));
  });

  it("should render correctly", async () => {
    const history = createMemoryHistory();
    const { queryByText } = render(
      <Router history={history}>
        <ImportUacPage />
      </Router>
    );

    expect(queryByText(/Import UACs from file/i)).toBeInTheDocument();
  });

  it("Select uac file - should display an error message if you dont select a file", async () => {
    await NavigateToSelectFile();

    await fireEvent.click(screen.getByText(/Continue/));

    await waitFor(() => {
      expect(screen.queryAllByText("Select a file")).toHaveLength(2);
    });
  });

  it("Select uac file - should display an error message if the selected file is not a .csv", async () => {
    await NavigateToSelectFileAndUpload("txt");

    await fireEvent.click(screen.getByText(/Continue/));

    await waitFor(() => {
      expect(screen.queryAllByText("File must be a .csv")).toHaveLength(2);
    });
  });

  it("Select uac file - should display an upload filed message if file fails to upload", async () => {
    mock.onPost("/api/v1/uac/import").reply(500);

    await NavigateToSelectFileAndUpload("csv");

    await waitFor(() => {
      expect(screen.queryAllByText("File upload failed")).toHaveLength(1);
    });
  });

  it("Select uac file - should navigate to the success page with uac_import count", async () => {
    mock.onPost("/api/v1/uac/import").reply(200, {uacs_imported: 12});

    await NavigateToSelectFileAndUpload("csv");

    await waitFor(() => {
      expect(screen.queryAllByText("UACs Imported: 12")).toHaveLength(1);
    });
  });


  it("Select uac file - should navigate to the success page with uac_import count - when the api doesn't respond with a count", async () => {
    mock.onPost("/api/v1/uac/import").reply(200);

    await NavigateToSelectFileAndUpload("csv");

    await waitFor(() => {
      expect(screen.queryAllByText("UACs Imported: 0")).toHaveLength(1);
    });
  });

  it(("Select uac file - should return a generic error when the import fails"), async () => {
    mock.onPost("/api/v1/uac/import").reply(500);

    await NavigateToSelectFileAndUpload("csv");

    await waitFor(() => {
      expect(screen.queryAllByText("File upload failed")).toHaveLength(1);
    });
  });



  it(("Select uac file - should return a specifc error when the import fails with a reason"), async () => {
    mock.onPost("/api/v1/uac/import").reply(500, {error: "All the kittens melted"});

    await NavigateToSelectFileAndUpload("csv");

    await waitFor(() => {
      expect(screen.queryAllByText("All the kittens melted")).toHaveLength(1);
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
    cleanup();
  });
});


async function NavigateToSelectFile() {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <ImportUacPage />
    </Router>
  );

  await waitFor(() => {
    expect(screen.queryAllByText("Select a uac file")).toHaveLength(1);
  });
}

async function NavigateToSelectFileAndUpload(fileExtension: string) {
  await NavigateToSelectFile();
  const inputEl = screen.getByLabelText(/Select a uac file/i);

  const file = new File(["(⌐□_□)"], `OPN2004A.${fileExtension}`, {
    type: fileExtension
  });

  Object.defineProperty(inputEl, "files", {
    value: [file]
  });

  fireEvent.change(inputEl);

  await fireEvent.click(screen.getByText(/Continue/));
}
