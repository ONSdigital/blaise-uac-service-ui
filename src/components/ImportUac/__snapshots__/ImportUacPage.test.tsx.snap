// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import UAC Page import uac page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="ons-page__main ons-u-mt-no"
        id="main-content"
      >
        <ul
          class="ons-list ons-list--bare ons-list--inline ons-u-mt-m"
        >
          <li
            class="ons-list__item"
          >
            <a
              href="/app"
              id="generate-uac-link"
            >
              Home
            </a>
          </li>
        </ul>
        <form
          action="#"
          id="formID"
        >
          <div>
            <div
              class="ons-u-mt-m"
            >
              <div
                class="ons-grid"
              >
                <div
                  class="ons-grid__col ons-col-8@m"
                >
                  <h1>
                    Import UACs from file
                  </h1>
                  <div
                    class="ons-panel ons-panel--info ons-panel--no-title  ons-u-mt-m"
                  >
                    <div
                      class="ons-panel__body "
                    >
                      <p>
                        The UAC file must be a Comma-Separated Values (CSV) file containing the column 
                        <em
                          class="ons-highlight"
                        >
                          Full_UAC
                        </em>
                        .
                      </p>
                    </div>
                  </div>
                  <div
                    class="ons-field"
                  >
                    <label
                      class="ons-label ons-label--with-description"
                      for="uac-selector"
                    >
                      Select a uac file
                    </label>
                    <span
                      class="ons-label__description  ons-input--with-description"
                      id="description-hint"
                    >
                      File type accepted is .csv
                    </span>
                    <input
                      accept=".csv"
                      class="ons-input ons-input--text ons-input-type__input ons-input--upload"
                      id="uac-selector"
                      name="Select a uac file"
                      props="[object Object]"
                      type="file"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ons-btn-group ons-u-mt-m"
            >
              <button
                class="ons-btn "
                data-testid="button"
                id="next-screen-button"
                type="submit"
              >
                <span
                  class="ons-btn__inner"
                >
                  Continue
                </span>
              </button>
            </div>
          </div>
        </form>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="ons-page__main ons-u-mt-no"
      id="main-content"
    >
      <ul
        class="ons-list ons-list--bare ons-list--inline ons-u-mt-m"
      >
        <li
          class="ons-list__item"
        >
          <a
            href="/app"
            id="generate-uac-link"
          >
            Home
          </a>
        </li>
      </ul>
      <form
        action="#"
        id="formID"
      >
        <div>
          <div
            class="ons-u-mt-m"
          >
            <div
              class="ons-grid"
            >
              <div
                class="ons-grid__col ons-col-8@m"
              >
                <h1>
                  Import UACs from file
                </h1>
                <div
                  class="ons-panel ons-panel--info ons-panel--no-title  ons-u-mt-m"
                >
                  <div
                    class="ons-panel__body "
                  >
                    <p>
                      The UAC file must be a Comma-Separated Values (CSV) file containing the column 
                      <em
                        class="ons-highlight"
                      >
                        Full_UAC
                      </em>
                      .
                    </p>
                  </div>
                </div>
                <div
                  class="ons-field"
                >
                  <label
                    class="ons-label ons-label--with-description"
                    for="uac-selector"
                  >
                    Select a uac file
                  </label>
                  <span
                    class="ons-label__description  ons-input--with-description"
                    id="description-hint"
                  >
                    File type accepted is .csv
                  </span>
                  <input
                    accept=".csv"
                    class="ons-input ons-input--text ons-input-type__input ons-input--upload"
                    id="uac-selector"
                    name="Select a uac file"
                    props="[object Object]"
                    type="file"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ons-btn-group ons-u-mt-m"
          >
            <button
              class="ons-btn "
              data-testid="button"
              id="next-screen-button"
              type="submit"
            >
              <span
                class="ons-btn__inner"
              >
                Continue
              </span>
            </button>
          </div>
        </div>
      </form>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
